From 58e300b95dd50b8f43106ac51c116b929f236039 Mon Sep 17 00:00:00 2001
From: Frank Bausch <francis.bausch@nrl.navy.mil>
Date: Fri, 28 Feb 2025 17:25:07 +0000
Subject: [PATCH 02/29] simVis: protect against access of null terrain engine
 (SIM-18009)

**JIRA Issue:** SIM-18009

**Description:** examples can segfault, especially on linux due to attempts to access null terrain engine

**Notes:** all SIMDIS-X accesses already protected.

**Testing Performed:** can no longer reproduce crashes on linux
---
 Examples/Ocean/Ocean.cpp                 |  9 ++++++++-
 Examples/SimpleServer/InstallOcean.cpp   |  2 +-
 SDK/simUtil/MousePositionManipulator.cpp | 10 ++++++++++
 SDK/simVis/Projector.cpp                 |  3 ++-
 4 files changed, 21 insertions(+), 3 deletions(-)

diff --git a/Examples/Ocean/Ocean.cpp b/Examples/Ocean/Ocean.cpp
index 500d7977..5c933276 100644
--- a/Examples/Ocean/Ocean.cpp
+++ b/Examples/Ocean/Ocean.cpp
@@ -1250,6 +1250,13 @@ int main(int argc, char** argv)
   viewer->setMap(map.get());
   osg::ref_ptr<simVis::SceneManager> scene = viewer->getSceneManager();
 
+  if (!scene->getMapNode() || !scene->getMapNode()->getTerrainEngine())
+  {
+    // this example can't even limp along without a terrain engine.
+    SIM_ERROR << "Ocean example cannot continue with no terrain engine.\n";
+    return 1;
+  }
+
   // the data store houses the entity data model:
   simData::MemoryDataStore dataStore;
   simVis::ScenarioDataStoreAdapter adapter(&dataStore, scene->getScenario());
@@ -1289,7 +1296,7 @@ int main(int argc, char** argv)
   }
 
   // if we're using Triton, install a module to "sink" the MSL=0 terrain down, creating makeshift bathymetry
-  if (bathymetryOffset != 0.0f)
+  if (bathymetryOffset != 0.0f && scene->getMapNode()->getTerrainEngine())
   {
     SIM_NOTICE << "Bathymetry offset = " << -bathymetryOffset << "\n";
     simVis::BathymetryGenerator* bgen = new simVis::BathymetryGenerator();
diff --git a/Examples/SimpleServer/InstallOcean.cpp b/Examples/SimpleServer/InstallOcean.cpp
index c23218f5..2841143f 100644
--- a/Examples/SimpleServer/InstallOcean.cpp
+++ b/Examples/SimpleServer/InstallOcean.cpp
@@ -108,7 +108,7 @@ void InstallOcean::install(simVis::SceneManager& scene)
     return;
 
   // Install the bathymetry offset
-  if (bathymetryOffset_ != 0.0)
+  if (bathymetryOffset_ != 0.0 && scene.getMapNode()->getTerrainEngine())
   {
     simVis::BathymetryGenerator* bathGen = new simVis::BathymetryGenerator();
     bathGen->setOffset(-bathymetryOffset_);
diff --git a/SDK/simUtil/MousePositionManipulator.cpp b/SDK/simUtil/MousePositionManipulator.cpp
index b9f125fc..874a9ce7 100644
--- a/SDK/simUtil/MousePositionManipulator.cpp
+++ b/SDK/simUtil/MousePositionManipulator.cpp
@@ -86,6 +86,11 @@ MousePositionManipulator::MousePositionManipulator(osgEarth::MapNode* mapNode, o
     scene_(scene)
 {
   assert(mapNode != nullptr);
+  if (!mapNode_->getTerrainEngine())
+  {
+    terrainEngineNode_ = nullptr;
+    return;
+  }
 #if OSGEARTH_SOVERSION >= 104
   terrainEngineNode_ = mapNode_->getTerrainEngine()->getNode();
 #else
@@ -122,6 +127,11 @@ void MousePositionManipulator::setMapNode(osgEarth::MapNode* mapNode)
     terrainEngineNode_ = nullptr;
     return;
   }
+  if (!mapNode_->getTerrainEngine())
+  {
+    terrainEngineNode_ = nullptr;
+    return;
+  }
 
 #if OSGEARTH_SOVERSION >= 104
   terrainEngineNode_ = mapNode_->getTerrainEngine()->getNode();
diff --git a/SDK/simVis/Projector.cpp b/SDK/simVis/Projector.cpp
index 59178701..60d3411b 100644
--- a/SDK/simVis/Projector.cpp
+++ b/SDK/simVis/Projector.cpp
@@ -1096,7 +1096,8 @@ void ProjectorNode::setMapNode(osgEarth::MapNode* mapNode)
     shadowCam_->removeChildren(0, shadowCam_->getNumChildren());
     if (mapNode)
     {
-      shadowCam_->addChild(mapNode->getTerrainEngine()->getNode());
+      if (mapNode->getTerrainEngine())
+        shadowCam_->addChild(mapNode->getTerrainEngine()->getNode());
       shadowCam_->setRenderingCache(nullptr);
     }
   }
-- 
2.47.1.windows.2

